server/extension.go

```diff
@@ -0,0 +1,155 @@
+	querypb "github.com/dolthub/vitess/go/vt/proto/query"
+	ast "github.com/dolthub/vitess/go/vt/sqlparser"
+	"sort"
+)
+
+func Intercept(h Interceptor) {
+	inters = append(inters, h)
+	sort.Slice(inters, func(i, j int) bool { return inters[i].Priority() < inters[j].Priority() })
+}
+
+var inters []Interceptor
+
+func withChain(h mysql.Handler) mysql.Handler {
+	var last Chain = h
+	for i := len(inters) - 1; i >= 0; i-- {
+		filter := inters[i]
+		next := last
+		last = &chainInterceptor{i: filter, c: next}
+	}
+	return &interceptorHandler{h: h, c: last}
+}
+
+type Interceptor interface {
+
+	// Priority returns the priority of the interceptor.
+	Priority() int
+
+	// Query is called when a connection receives a query.
+	// Note the contents of the query slice may change after
+	// the first call to callback. So the Handler should not
+	// hang on to the byte slice.
+	Query(chain Chain, c *mysql.Conn, query string, callback func(res *sqltypes.Result, more bool) error) error
+
+	// MultiQuery is called when a connection receives a query and the
+	// client supports MULTI_STATEMENT. It should process the first
+	// statement in |query| and return the remainder. It will be called
+	// multiple times until the remainder is |""|.
+	MultiQuery(chain Chain, c *mysql.Conn, query string, callback func(res *sqltypes.Result, more bool) error) (string, error)
+
+	// Prepare is called when a connection receives a prepared
+	// statement query.
+	Prepare(chain Chain, c *mysql.Conn, query string) ([]*querypb.Field, error)
+
+	// StmtExecute is called when a connection receives a statement
+	// execute query.
+	StmtExecute(chain Chain, c *mysql.Conn, prepare *mysql.PrepareData, callback func(*sqltypes.Result) error) error
+}
+
+type Chain interface {
+
+	// ComQuery is called when a connection receives a query.
+	// Note the contents of the query slice may change after
+	// the first call to callback. So the Handler should not
+	// hang on to the byte slice.
+	ComQuery(c *mysql.Conn, query string, callback func(res *sqltypes.Result, more bool) error) error
+
+	// ComMultiQuery is called when a connection receives a query and the
+	// client supports MULTI_STATEMENT. It should process the first
+	// statement in |query| and return the remainder. It will be called
+	// multiple times until the remainder is |""|.
+	ComMultiQuery(c *mysql.Conn, query string, callback func(res *sqltypes.Result, more bool) error) (string, error)
+
+	// ComPrepare is called when a connection receives a prepared
+	// statement query.
+	ComPrepare(c *mysql.Conn, query string) ([]*querypb.Field, error)
+
+	// ComStmtExecute is called when a connection receives a statement
+	// execute query.
+	ComStmtExecute(c *mysql.Conn, prepare *mysql.PrepareData, callback func(*sqltypes.Result) error) error
+}
+
+type chainInterceptor struct {
+	i Interceptor
+	c Chain
+}
+
+func (that *chainInterceptor) ComQuery(c *mysql.Conn, query string, callback func(res *sqltypes.Result, more bool) error) error {
```

Comment:
Name the receiver `ci` to match package style

Reply:
Done